File: usaco/gold/fencedin.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using edge = tuple<int, int, int>;

struct DSU {
    int count;
    vector<int> size, parent;
    DSU(int n) : count(n), parent(n, -1), size(n, 1) {}
    void unite(int a, int b) {
        a = find(a), b = find(b);
        if (a == b) return;
        if (size[a] < size[b]) swap(a,b);
        parent[b] = a;
        size[a] += size[b];
        --count;
    }
    int find(int a) {
        if (parent[a] == -1) return a;
        return parent[a] = find(parent[a]);
    }
};

int main() {
    freopen("fencedin.in", "r", stdin);
    freopen("fencedin.out", "w", stdout);
    ll ans = 0;
    int A, B, n, m;
    cin >> A >> B >> n >> m;
    DSU uf((n+1)*(m+1));
    vector<int> vert(n+2), hori(m+2);
    vert[0] = 0, vert[n+1] = A;
    hori[0] = 0, hori[m+1] = B;
    for (int i = 0; i < n; ++i) cin >> vert[i+1];
    for (int i = 0; i < m; ++i) cin >> hori[i+1];
    sort(begin(vert), end(vert));
    sort(begin(hori), end(hori));
    vector<edge> edges;
    int dx = n+1;
    for (int i = 0; i < n+1; ++i) {
        int w = vert[i+1] - vert[i];
        for (int j = 0; j < m; ++j) {
            int a = j * dx + i;
            int b = (j+1) * dx + i;
            edges.push_back( {w,a,b} );
        }
    }
    for (int j = 0; j < m+1; ++j) {
        int w = hori[j+1] - hori[j];
        for (int i = 0; i < n; ++i) {
            int a = j * dx + i;
            int b = a + 1;
            edges.push_back( {w,a,b} );
        }
    }
    sort(begin(edges), end(edges));
    for (auto &e : edges) {
        auto [w,a,b] = e;
        if (uf.find(a) == uf.find(b))
            continue;
        uf.unite(a, b);
        ans += w;
        if (uf.count == 1)
            break;
    }
    cout << ans << endl;

}

File: usaco/gold/fencedin.py
import sys
from collections import defaultdict
intput = lambda : map(int, input().split())

def union(a, b):
    global count
    a, b = find(a), find(b)
    if a == b: return False
    if size[a] < size[b]: a,b = b,a
    parent[b] = a
    size[a] += size[b]
    count -= 1
    return True

def find(a):
    if parent[a] == -1: return a
    parent[a] = find(parent[a])
    return parent[a]

if __name__ == '__main__':
    sys.stdin  = open('fencedin.in',  'r')
    sys.stdout = open('fencedin.out', 'w')
    ans = 0
    A, B, n, m = intput()
    count  = (n+1) * (m+1)
    size   = defaultdict(lambda : 1)
    parent = defaultdict(lambda : -1)
    vert = [0] + sorted( int(input()) for _ in range(n) ) + [A]
    hori = [0] + sorted( int(input()) for _ in range(m) ) + [B]
    edges = []
    for i in range(n+1):
        w = vert[i + 1] - vert[i]
        for j in range(m):
            a, b = (j,i), (j+1,i)
            edges.append( (w,a,b) )
    for j in range(m+1):
        w = hori[j + 1] - hori[j]
        for i in range(n):
            a, b = (j,i), (j,i+1)
            edges.append( (w,a,b) )
    edges.sort()
    for e in edges:
        w, a, b = e
        if find(a) == find(b):
            continue
        union(a, b)
        ans += w
        if count == 1:
            break

    print(ans)
